Tests/TSFCASUtilitiesTests/LinkedListStreamTests.swift

```diff
@@ -0,0 +1,184 @@
+        }
+
+        let reader = LLBLinkedListStreamReader(db)
+
+        let latestID = try writer.latestID!.wait()
+
+        var readStream = [String]()
+
+        try reader.read(id: latestID, channels: [0, 1], ctx) { (channel, data) -> Bool in
+            guard let block = String(data: Data(data.readableBytesView), encoding: .utf8) else {
+                XCTFail("Could not read string from stream")
+                return false
+            }
+            readStream.append(block)
+            return true
+        }.wait()
+
+        let filteredWriteStream = writeStream.filter { $0.0 <= 1 }.map { $0.1 }
+
+        XCTAssertEqual(readStream, filteredWriteStream)
+    }
+
+    func testStreamReadLimit() throws {
+        let db = LLBInMemoryCASDatabase(group: group)
+        let ctx = Context()
+
+        let writeStream = Array(0...20).map { "Stream line \($0)" }
+
+        var writer = LLBLinkedListStreamWriter(db)
+
+        for block in writeStream {
+            writer.append(data: LLBByteBuffer(string: block), ctx)
+        }
+
+        let reader = LLBLinkedListStreamReader(db)
+
+        let latestID = try writer.latestID!.wait()
+
+        var readStream = [String]()
+
+        var stopped = false
+        try reader.read(id: latestID, ctx) { (channel, data) -> Bool in
+            // Read only 5 elements
+            if readStream.count == 5 {
+                stopped = true
+                return false
+            }
+            guard !stopped else {
+                XCTFail("Requested to stop but kept receiving data")
+                return false
+            }
+
+            guard let block = String(data: Data(data.readableBytesView), encoding: .utf8) else {
+                XCTFail("Could not read string from stream")
+                return false
+            }
+            readStream.append(block)
+            return true
+        }.wait()
+
+        XCTAssertEqual(readStream, Array(writeStream.prefix(5)))
+    }
+
+
+    func testStreamFromPreviousState() throws {
+        let db = LLBInMemoryCASDatabase(group: group)
+        let ctx = Context()
+
+        let writeStream = Array(0...20).map { "Stream line \($0)" }
+
+        var writer = LLBLinkedListStreamWriter(db)
+
+        for block in writeStream {
+            writer.append(data: LLBByteBuffer.init(string: block), ctx)
+        }
+
+        let startMarker = try writer.latestID!.wait()
+
+        let writeStream2 = Array(21...40).map { "Stream line \($0)" }
+
+        for block in writeStream2 {
+            writer.append(data: LLBByteBuffer.init(string: block), ctx)
+        }
+
+        let reader = LLBLinkedListStreamReader(db)
+
+        let latestID = try writer.latestID!.wait()
+
+        var readStream = [String]()
+
+        try reader.read(id: latestID, lastReadID: startMarker, ctx) { (channel, data) -> Bool in
+            guard let block = String(data: Data(data.readableBytesView), encoding: .utf8) else {
```

Comment:
nit: `String(decoding: ..., as: UTF8.self)` is a non-failable version of ^^, could be visually simpler.

Reply:
nifty :) keep forgetting about that one